### PR TITLE
[GCS] refactor the GCS Client Dynamic Resource Interface

### PR DESCRIPTION
**Pull request by micafan**
Maybe we can move dynamic resource to NodeStateAccessor

## Checks

- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [X] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

**Comment by micafan** [comment]
Still working on UT. And not change raylet usage yet.

**Comment by micafan** [comment]
> Still working on UT. And not change raylet usage yet.

Finished. Waiting for review.

**Review by zhijunfu — APPROVED** [review]

File: src/ray/gcs/accessor.h [hunk 1]
```diff
@@ -352,6 +352,45 @@ class NodeInfoAccessor {
...
+  /// Update dynamic resources of node in GCS asynchronously.
+  ///
+  /// \param node_id The ID of node to update dynamic resources.
+  /// \param resources The dynamic resources of node to be updated.
+  /// \param callback Callback that will be called after update finishes.
+  virtual Status AsyncUpdateResource(const ClientID &node_id,
+                                     const ResourceMap &resources,
+                                     const StatusCallback &callback) = 0;
+
+  /// Delete resources of an node from GCS asynchronously.
```
Comment (ffbin) [inline src/ray/gcs/accessor.h]: a node

File: src/ray/gcs/accessor.h [hunk 2]
```diff
@@ -352,6 +352,45 @@ class NodeInfoAccessor {
...
+  // TODO(micafan) Define ResourceMap in GCS proto.
+  typedef std::unordered_map<std::string, std::shared_ptr<rpc::ResourceTableData>>
+      ResourceMap;
+
+  /// Get node's dynamic resources from GCS asynchronously.
+  ///
+  /// \param node_id The ID of node to lookup dynamic resources.
+  /// \param callback Callback that will be called after lookup finishes.
+  /// \return Status
+  virtual Status AsyncGetResource(const ClientID &node_id,
```
Comment (raulchen) [inline src/ray/gcs/accessor.h]:
```suggestion
  virtual Status AsyncGetResources(const ClientID &node_id,
```

File: src/ray/gcs/accessor.h [hunk 3]
```diff
@@ -352,6 +352,45 @@ class NodeInfoAccessor {
...
+                                  const OptionalItemCallback<ResourceMap> &callback) = 0;
+
+  /// Update dynamic resources of node in GCS asynchronously.
+  ///
+  /// \param node_id The ID of node to update dynamic resources.
+  /// \param resources The dynamic resources of node to be updated.
+  /// \param callback Callback that will be called after update finishes.
+  virtual Status AsyncUpdateResource(const ClientID &node_id,
```
Comment (raulchen) [inline src/ray/gcs/accessor.h]:
```suggestion
  virtual Status AsyncUpdateResources(const ClientID &node_id,
```

File: src/ray/gcs/accessor.h [hunk 4]
```diff
@@ -352,6 +352,45 @@ class NodeInfoAccessor {
...
+                                     const ResourceMap &resources,
+                                     const StatusCallback &callback) = 0;
+
+  /// Delete resources of an node from GCS asynchronously.
+  ///
+  /// \param node_id The ID of node to delete resources from GCS.
+  /// \param resource_names The names of resource to be deleted.
+  /// \param callback Callback that will be called after delete finishes.
+  virtual Status AsyncDeleteResource(const ClientID &node_id,
```
Comment (raulchen) [inline src/ray/gcs/accessor.h]:
```suggestion
  virtual Status AsyncDeleteResources(const ClientID &node_id,
```

File: src/ray/gcs/accessor.h [hunk 5]
```diff
@@ -352,6 +352,45 @@ class NodeInfoAccessor {
...
+                                     const std::vector<std::string> &resource_names,
+                                     const StatusCallback &callback) = 0;
+
+  /// Subscribe to any update operations of dynamic resources.
+  ///
+  /// \param subscribe Callback that will be called when any resource is updated.
+  /// \param done Callback that will be called when subscription is complete.
+  /// \return Status
+  virtual Status AsyncSubscribeResource(
```
Comment (raulchen) [inline src/ray/gcs/accessor.h]:
```suggestion
  virtual Status AsyncSubscribeToResources(
```

**Review by raulchen — APPROVED** [review]
LGTM, only a few comments regarding naming.

File: src/ray/gcs/accessor.h [hunk 6]
```diff
@@ -321,6 +321,45 @@ class NodeInfoAccessor {
   /// \return Whether the node is removed.
   virtual bool IsRemoved(const ClientID &node_id) const = 0;
 
+  // TODO(micafan) Define ResourceMap in GCS proto.
+  typedef std::unordered_map<std::string, std::shared_ptr<rpc::ResourceTableData>>
+      ResourceMap;
+
+  /// Get node's dynamic resources from GCS asynchronously.
```
[the remark this answers is not in the document]
Comment (micafan) [reply src/ray/gcs/accessor.h]: done

File: src/ray/gcs/accessor.h [hunk 7]
```diff
@@ -321,6 +321,45 @@ class NodeInfoAccessor {
...
+  ///
+  /// \param node_id The ID of node to lookup dynamic resources.
+  /// \param callback Callback that will be called after lookup finishes.
+  /// \return Status
+  virtual Status AsyncGetResource(const ClientID &node_id,
+                                  const OptionalItemCallback<ResourceMap> &callback) = 0;
+
+  /// Update dynamic resources of node in GCS asynchronously.
```
[the remark this answers is not in the document]
Comment (micafan) [reply src/ray/gcs/accessor.h]: done

File: src/ray/gcs/accessor.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (micafan) [reply src/ray/gcs/accessor.h]: done

File: src/ray/gcs/accessor.h (hunk 2 above)
Comment (micafan) [reply src/ray/gcs/accessor.h]: done

Comment (micafan) [reply src/ray/gcs/accessor.h]: done

File: src/ray/gcs/accessor.h (hunk 3 above)
Comment (micafan) [reply src/ray/gcs/accessor.h]: done

File: src/ray/gcs/accessor.h [hunk 8]
```diff
@@ -321,6 +321,45 @@ class NodeInfoAccessor {
...
+  ///
+  /// \param node_id The ID of node to update dynamic resources.
+  /// \param resources The dynamic resources of node to be updated.
+  /// \param callback Callback that will be called after update finishes.
+  virtual Status AsyncUpdateResource(const ClientID &node_id,
+                                     const ResourceMap &resources,
+                                     const StatusCallback &callback) = 0;
+
+  /// Delete resources of an node from GCS asynchronously.
```
[the remark this answers is not in the document]
Comment (micafan) [reply src/ray/gcs/accessor.h]: done

File: src/ray/gcs/accessor.h (hunk 1 above)
Comment (micafan) [reply src/ray/gcs/accessor.h]: done

File: src/ray/gcs/accessor.h (hunk 4 above)
Comment (micafan) [reply src/ray/gcs/accessor.h]: done

File: src/ray/gcs/accessor.h [hunk 9]
```diff
@@ -321,6 +321,45 @@ class NodeInfoAccessor {
...
+  ///
+  /// \param node_id The ID of node to delete resources from GCS.
+  /// \param resource_names The names of resource to be deleted.
+  /// \param callback Callback that will be called after delete finishes.
+  virtual Status AsyncDeleteResource(const ClientID &node_id,
+                                     const std::vector<std::string> &resource_names,
+                                     const StatusCallback &callback) = 0;
+
+  /// Subscribe to any update operations of dynamic resources.
```
[the remark this answers is not in the document]
Comment (micafan) [reply src/ray/gcs/accessor.h]: done

File: src/ray/gcs/accessor.h (hunk 5 above)
Comment (micafan) [reply src/ray/gcs/accessor.h]: done

**Pull request merged**